File: 15-1.rb
#!/usr/bin/env ruby

require 'numo/narray'

class Map
  def initialize(input, row)
    sensor_regexp = /^Sensor at x=(?<sx>[[:digit:]-]+), y=(?<sy>[[:digit:]-]+): closest beacon is at x=(?<bx>[[:digit:]-]+), y=(?<by>[[:digit:]-]+)$/
    @row = row

    min_x = 2**30 - 1
    max_x = 0
    max_dist = 0
    sensors = []
    beacons = []

    input.each do |sensor|
      m = sensor_regexp.match(sensor)
      sx = m['sx'].to_i
      sy = m['sy'].to_i
      bx = m['bx'].to_i
      by = m['by'].to_i
      min_x = [min_x, sx, bx].min
      max_x = [max_x, sx, bx].max
      max_dist = [max_dist, (sx - bx).abs + (sy - by).abs + 1].max
      sensors.append [sx, sy]
      beacons.append [bx, by]
    end

    @width = (max_x - min_x) + 2 * max_dist + 1
    @x_offset = min_x - max_dist
    @map = Numo::UInt8.zeros(@width)

    sensors.size.times do |index|
      sx = sensors[index][0] - @x_offset
      sy = sensors[index][1]
      bx = beacons[index][0] - @x_offset
      by = beacons[index][1]
      @map[sx] = 1 if sy == @row
      @map[bx] = 2 if by == @row

      manhattan_distance = (sx - bx).abs + (sy - by).abs
      next unless (sy - @row).abs < manhattan_distance

      num_cols = manhattan_distance - (sy - @row).abs + 1
      num_cols.times do |x_dist|
        @map[sx - x_dist] = 3 if sx - x_dist >= 0 and @map[sx - x_dist].zero?
        @map[sx + x_dist] = 3 if sx + x_dist < @width and @map[sx + x_dist].zero?
      end
    end
  end

  def count_impossibles
    count = 0
    @map.each { |cell| count += 1 if [1, 3].include? cell }
    count
  end

  def to_s
    s = "<#{self.class}:\n"
    s += "Row: #{@row}, x offset: #{@x_offset}, width: #{@width}\n"
    @width.times do |col|
      s += case @map[col]
           when 0
             '.'
           when 1
             'S'
           when 2
             'B'
           when 3
             '#'
           end
    end
    s += "\n>"
    s
  end

  alias inspect to_s
end

input = File.read('15.input').lines.map(&:strip)

map = Map.new(input, 2_000_000)

print map.count_impossibles, "\n"

File: 15-1.py
#!/usr/bin/env python

import numpy
import re

class Map:
    def __init__(self, input, row):
        sensor_regexp = re.compile("^Sensor at x=(?P<sx>[0-9-]+), y=(?P<sy>[0-9-]+): closest beacon is at x=(?P<bx>[0-9-]+), y=(?P<by>[0-9-]+)$")
        self.row = row

        min_x = pow(2, 30) - 1
        max_x = 0
        max_dist = 0
        sensors = []
        beacons = []

        for sensor in input:
            m = sensor_regexp.match(sensor)
            sx = int(m['sx'])
            sy = int(m['sy'])
            bx = int(m['bx'])
            by = int(m['by'])
            min_x = min(min_x, sx, bx)
            max_x = max(max_x, sx, bx)
            max_dist = max(max_dist, abs(sx - bx) + abs(sy - by) + 1)
            sensors.append((sx, sy))
            beacons.append((bx, by))

        self.width = (max_x - min_x) + 2 * max_dist + 1
        self.x_offset = min_x - max_dist
        self.map = numpy.zeros((self.width), dtype=numpy.uint)

        for index in range(len(sensors)):
            sx = sensors[index][0] - self.x_offset
            sy = sensors[index][1]
            bx = beacons[index][0] - self.x_offset
            by = beacons[index][1]
            if sy == self.row:
                self.map[sx] = 1
            if by == self.row:
                self.map[bx] = 2

            manhattan_distance = abs(sx - bx) + abs(sy - by)
            if not abs(sy - self.row) < manhattan_distance:
                continue

            num_cols = manhattan_distance - abs(sy - self.row) + 1
            for x_dist in range(num_cols):
                if sx - x_dist >= 0 and self.map[sx - x_dist] == 0:
                    self.map[sx - x_dist] = 3
                if sx + x_dist < self.width and self.map[sx + x_dist] == 0:
                    self.map[sx + x_dist] = 3

    def count_impossibles(self):
        count = 0
        for cell in self.map:
            if cell == 1 or cell == 3:
                count += 1
        return count

    def __str__(self):
        s = f"<{self.__class__.__name__}:\n"
        s += f"Row: #{self.row}, x offset: #{self.x_offset}, width: #{self.width}\n"
        for col in range(self.width):
            if self.map[col] == 0:
                s += "."
            elif self.map[col] == 1:
                s += "S"
            elif self.map[col] == 2:
                s += "B"
            elif self.map[col] == 3:
                s += "#"
        s += "\n>"
        return s


lines = [line.strip() for line in open('15.input').readlines()]

obj = Map(lines, 2_000_000)
print(obj.count_impossibles())
